src/sesclient.native/sesclient.native.cpp

```diff
@@ -43,62 +49,172 @@ struct Initializer
+        && !parameters.contains("--common-name"))
+    {
+        // We have neither value - and that's ok
+        return true;
+    }
+
+    if (!parameters.contains("--thumbprint"))
+    {
+        std::cerr << "--thumbprint must also be used when --common-name is used" << std::endl;
+        return false;
+    }
+
+    if (!parameters.contains("--common-name"))
+    {
+        std::cerr << "--common-name must also be used when --thumbprint is used" << std::endl;
+        return false;
+    }
+
+    auto thumbprintParameter = parameters.find("--thumbprint");
+    auto commonNameParameter = parameters.find("--common-name");
+
+    Microsoft::Azure::Batch::SoftwareEntitlement::AddSslCertificate(thumbprintParameter, commonNameParameter);
+    return true;
+}
+
+std::string readToken(ParameterParser& parameters)
+{
+    auto token = parameters.find("--token");
+    if (token == "-")
+    {
+        // Read the token from stdin.
+        std::getline(std::cin, token);
+    }
+
+    return token;
+}
 
 int main(int argc, char** argv)
 {
     try
     {
         Initializer init;
+        ParameterParser parser;
+
+        parser.parse(argc, argv);
 
-        if (argc != 11)
+        if (parser.shouldShowUsage())
         {
             ShowUsage(argv[0]);
             return -1;
         }
 
-        std::unordered_map<std::string, std::string> parameters;
-        while (argc != 1)
+        if (parser.hasConfigurationError())
         {
-            auto value = argv[--argc];
-            auto key = argv[--argc];
-            parameters.emplace(key, value);
+            return -EINVAL;
         }
 
-        for (const auto& param : parameterNames)
-        {
-            if (parameters.find(param) == parameters.end())
-            {
-                std::cout << "Missing parameter " << param << std::endl;
-                return -EINVAL;
-            }
-        }
+        auto token = readToken(parser);
+        auto connectionConfigured = configureConnection(parser);
 
-        std::string token = parameters.find("--token")->second;
-        if (token == "-")
+        if (!connectionConfigured)
         {
-            //
-            // Read the token from stdin.
-            //
-            std::getline(std::cin, token);
+            return -EINVAL;
         }
 
-        Microsoft::Azure::Batch::SoftwareEntitlement::AddSslCertificate(
-            parameters.find("--thumbprint")->second,
-            parameters.find("--common-name")->second
-        );
-
         auto entitlement = Microsoft::Azure::Batch::SoftwareEntitlement::GetEntitlement(
-            parameters.find("--url")->second,
+            parser.find("--url"),
             token,
-            parameters.find("--application")->second
+            parser.find("--application")
         );
+
         std::cout << entitlement->Id() << std::endl;
     }
     catch (const std::exception& e)
     {
-        std::cout << e.what() << std::endl;
+        std::cerr << e.what() << std::endl;
```

Comment:
It might be better to properly crash on unexpected errors, so you can get a dump